applications/StructuralMechanicsApplication/custom_response_functions/adjoint_elements/adjoint_finite_difference_truss_element_3D2N.cpp

```diff
@@ -0,0 +1,332 @@
+
+    CheckVariables();
+    CheckDofs();
+    CheckProperties(rCurrentProcessInfo);
+
+    KRATOS_ERROR_IF(this->GetGeometry().Length() < std::numeric_limits<double>::epsilon())
+        << "Element #" << this->Id() << " has a length of zero!" << std::endl;
+
+    return 0;
+
+    KRATOS_CATCH("")
+}
+
+
+void AdjointFiniteDifferenceTrussElement::CheckVariables()
+{
+    KRATOS_CHECK_VARIABLE_KEY(DISPLACEMENT);
+    KRATOS_CHECK_VARIABLE_KEY(VELOCITY);
+    KRATOS_CHECK_VARIABLE_KEY(ACCELERATION);
+    KRATOS_CHECK_VARIABLE_KEY(DENSITY);
+    KRATOS_CHECK_VARIABLE_KEY(CROSS_AREA);
+    KRATOS_CHECK_VARIABLE_KEY(ADJOINT_DISPLACEMENT);
+}
+
+void AdjointFiniteDifferenceTrussElement::CheckDofs()
+{
+    const SizeType number_of_nodes = this->GetGeometry().size();
+
+    for (IndexType i = 0; i < number_of_nodes; ++i)
+    {
+        const NodeType &r_node = this->GetGeometry()[i];
+
+        KRATOS_CHECK_VARIABLE_IN_NODAL_DATA(DISPLACEMENT, r_node);
+        KRATOS_CHECK_VARIABLE_IN_NODAL_DATA(ADJOINT_DISPLACEMENT, r_node);
+
+        KRATOS_CHECK_DOF_IN_NODE(ADJOINT_DISPLACEMENT_X, r_node);
+        KRATOS_CHECK_DOF_IN_NODE(ADJOINT_DISPLACEMENT_Y, r_node);
+        KRATOS_CHECK_DOF_IN_NODE(ADJOINT_DISPLACEMENT_Z, r_node);
+    }
+}
+
+void AdjointFiniteDifferenceTrussElement::CheckProperties(const ProcessInfo& rCurrentProcessInfo)
+{
+    const double numerical_limit = std::numeric_limits<double>::epsilon();
+    const PropertiesType & r_properties = GetProperties();
+
+    KRATOS_ERROR_IF(r_properties.Has(CROSS_AREA) == false || r_properties[CROSS_AREA] <= numerical_limit)
+    << "CROSS_AREA not provided for this element" << this->Id() << std::endl;
+    KRATOS_ERROR_IF(r_properties.Has(YOUNG_MODULUS) == false || r_properties[YOUNG_MODULUS] <= numerical_limit)
+    << "YOUNG_MODULUS not provided for this element" << this->Id() << std::endl;
+    KRATOS_ERROR_IF_NOT( r_properties.Has(DENSITY) )
+    << "DENSITY not provided for this element" << this->Id() << std::endl;
+
+    KRATOS_ERROR_IF_NOT(r_properties.Has(CONSTITUTIVE_LAW))
+    << "CONSTITUTIVE_LAW not provided for element " << this->Id() << std::endl;
+    const ConstitutiveLaw::Pointer& cl = r_properties[CONSTITUTIVE_LAW];
+    KRATOS_ERROR_IF(cl == nullptr)
+    << "CONSTITUTIVE_LAW not provided for element " << this->Id() << std::endl;
+    cl->Check(r_properties ,this->GetGeometry(),rCurrentProcessInfo);
+}
+
+double AdjointFiniteDifferenceTrussElement::GetPerturbationSizeModificationFactor(const Variable<array_1d<double,3>>& rDesignVariable)
+{
+    KRATOS_TRY;
+
+    if(rDesignVariable == SHAPE)
+        return this->CalculateReferenceLength();
+    else
+        return 1.0;
+
+    KRATOS_CATCH("")
+}
+
+double AdjointFiniteDifferenceTrussElement::CalculateReferenceLength()
+{
+    KRATOS_TRY;
+    const double numerical_limit = std::numeric_limits<double>::epsilon();
+    const double dx = this->GetGeometry()[1].X0() - this->GetGeometry()[0].X0();
```

Comment:
isn't this the `Length` function of the Geometry?

Reply:
Delivers the ```Length``` function always the initial length?